src/imitation/data/rollout.py

```diff
@@ -142,20 +148,26 @@ def add_steps_and_auto_finish(
             each `True` in the `dones` argument.
         """
         trajs: List[types.TrajectoryWithRew] = []
-        for env_idx in range(len(obs)):
+        wrapped_obs = types.DictObs.maybe_wrap(obs)
+
+        # len of dictobs is the shape[0] of each value array - which here is # of envs
+        for env_idx in range(len(wrapped_obs)):
             assert env_idx in self.partial_trajectories
             assert list(self.partial_trajectories[env_idx][0].keys()) == ["obs"], (
                 "Need to first initialize partial trajectory using "
                 "self._traj_accum.add_step({'obs': ob}, key=env_idx)"
             )
 
-        zip_iter = enumerate(zip(acts, obs, rews, dones, infos))
+        zip_iter = enumerate(zip(acts, wrapped_obs, rews, dones, infos))
         for env_idx, (act, ob, rew, done, info) in zip_iter:
             if done:
                 # When dones[i] from VecEnv.step() is True, obs[i] is the first
                 # observation following reset() of the ith VecEnv, and
                 # infos[i]["terminal_observation"] is the actual final observation.
                 real_ob = info["terminal_observation"]
+                if isinstance(real_ob, dict):
+                    # TODO: does this need to be unsqueezed or something?
+                    real_ob = types.DictObs(real_ob)
```

Comment:
I'm confused why this is being applied here (only to the terminal observation) rather than always (i.e. after the else statement on line 173, unindented).

Reply:
It's applied to all observations above (before we can iterate through the environments).